### ISSUE TITLE
Canvas.toDataURL('image/jpeg') of node canvas cannot be used because of a jsdom bug

### ISSUE THREAD

**Issue by challakoushik** [jsdom/jsdom#2199]
Basically, in HTMLCanvasElement.js line 53 only two arguments are being passed to Canvas.toDataURL.
```
HTMLCanvasElement.prototype.toDataURL = function toDataURL() {
  if (!this || !module.exports.is(this)) {
    throw new TypeError("Illegal invocation");
  }
  const args = [];
  {
    let curArg = arguments[0];
    if (curArg !== undefined) {
      curArg = conversions["DOMString"](curArg, {
        context: "Failed to execute 'toDataURL' on 'HTMLCanvasElement': parameter 1"
      });
    }
    args.push(curArg);
  }
  {
    let curArg = arguments[1];
    if (curArg !== undefined) {
      curArg = conversions["any"](curArg, {
        context: "Failed to execute 'toDataURL' on 'HTMLCanvasElement': parameter 2"
      });
    }
    args.push(curArg);
  }
  return this[impl].toDataURL(...args);
};
```

Whereas in node-canvas the toDataURL function needs 3 arguments in the case of 'image/jpeg' otherwise it throws the error : 'Missing required callback function for format "image/jpeg"'
```

if ('image/jpeg' === type) {
    if (undefined === fn) {
      throw new Error('Missing required callback function for format "image/jpeg"');
    }

    var stream = this.jpegStream(opts);
    // note that jpegStream is synchronous
    var buffers = [];
    stream.on('data', function (chunk) {
      buffers.push(chunk);
    });
    stream.on('error', function (err) {
      fn(err);
    });
    stream.on('end', function() {
      var result = 'data:image/jpeg;base64,' + Buffer.concat(buffers).toString('base64');
      fn(null, result);
    });
  }
```
This is really important to me because I am not able to create a pdf using jpg images generated through canvas using nodeJS. Would appreciate any help on this

**Comment by domenic** [comment]
There's no way to use toDataURL with three arguments in the spec, so jsdom will not implement any such version. If node-canvas does not support any two-argument toDataURL, then it's just not possible with jsdom.

**Comment by Sebmaster** [comment]
Reopening this because we probably either want to fall back to png, or throw the proper error in this case (not sure what the spec says on unsupported image types).

**Comment by FrecksterGIT** [comment]
Spec on this:
https://html.spec.whatwg.org/multipage/canvas.html#a-serialisation-of-the-bitmap-as-a-file

> User agents must support PNG ("image/png"). User agents may support other types. If the user agent does not support the requested type, then it must create the file using the PNG format.

So I guess jsdom doesn't support jpeg and should fallback to png here.

**Comment by odrzutowiec** [comment]
Fallback would be nice! I'm using Jest jsdom environment to test my react photo upload component with thumbnail previews done with ```img.src=foourl; img.onload(barfunc)```. I can't test it without jsdom and node-canvas playing nice together and being able to load my images.

**Comment by zbjornson** [comment]
The upstream canvas PR was merged but is 2.x only. I can fix this when I update #1964 (when canvas 2.x finally ships). (Feel free to assign the issue to me.)

**Comment by asturur** [comment]
this should be fixed now.
